apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,503 @@
+import concurrent.futures
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ContextManager,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+    cast,
+)
+
+import anndata
+import numpy as np
+import numpy.typing as npt
+import pandas as pd
+import pyarrow as pa
+from typing_extensions import Literal, TypedDict
+
+from .. import somacore  # to be replaced by somacore package, when available
+from ..dataframe import DataFrame as SOMADataFrame
+from ..sparse_nd_array import SparseNDArrayRead
+
+if TYPE_CHECKING:
+    from ..experiment import Experiment
+from ..sparse_nd_array import SparseNDArray as SOMASparseNDArray
+from .anndata import _make_anndata
+from .axis import AxisQuery
+from .eq_types import AxisColumnNames, ExperimentAxisQueryReadArrowResult
+
+AxisJoinIds = TypedDict(
+    "AxisJoinIds",
+    {
+        "obs": pa.Array,
+        "var": pa.Array,
+    },
+)
+
+MatrixAxisQuery = TypedDict(
+    "MatrixAxisQuery",
+    {
+        "obs": AxisQuery,
+        "var": AxisQuery,
+    },
+)
+
+
+class ExperimentAxisQuery(ContextManager["ExperimentAxisQuery"]):
+    """
+    ExperimentAxisQuery allows easy selection and extraction of data from a single soma.Measurement
+    in a soma.Experiment, by obs/var (axis) coordinates and/or value filter [lifecycle: experimental].
+
+    The primary use for this class is slicing Experiment ``X`` layers by obs or var value and/or coordinates.
+    Slicing on SparseNDArray ``X`` matrices is supported; DenseNDArray is not supported at this time.
+
+    IMPORTANT: this class is not thread-safe.
+
+    IMPORTANT: this query class assumes it can store the full result of both axis dataframe
+    queries in memory, and only provides incremental access to the underlying X NDArray. API
+    features such as `n_obs` and `n_vars` codify this in the API.
+
+    IMPORTANT: you must call `close()` on any instance of this class in order to release
+    underlying resources. The ExperimentAxisQuery is a context manager, and it is recommended
+    that you use the following pattern to make this easy and safe:
+    ```
+        with ExperimentAxisQuery(...) as query:
+            ...
+    ```
+    """
+
+    experiment: "Experiment"
+    measurement_name: str
+
+    _query: MatrixAxisQuery
```

Comment:
This has been confusing me a bit, just in terms of readability. _After_ reading a few times I get it, but my concern is how to make it more immediately graspable on a first read.

Below we have `query = self._query[axis]` ... this is confusing since query is singular, queries (which we're not saying) is plural, and I was at first inclined to suggest `query = self._queries[axis]`. But `self._query` isn't a (plural) queries object, it's a `MatrixAxisQuery` ...

So my suggestion is to rename `_query` to `_matrix_axis_query` and then below have `axis_query = self._matrix_axis_query[axis]`

Reply:
change made